futurewei-cloud/arktos
# per-network service IP allocation - part 2: implement method of getting service ip allocator per network

**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
This is part 2 for per-network service IP allocation, follwoing [part 1](https://github.com/futurewei-cloud/arktos/pull/484)

1. provides the method implementation of per network service IP allocation;
2. adds test code;

Per-network Service IP allocation is depicted in [service IP section, multi-tenant network design proposal](https://github.com/futurewei-cloud/arktos/blob/master/docs/design-proposals/multi-tenancy/multi-tenancy-network.md#service-ip).

**How to verify on local dev env**:
1. start etcd
2. start kube-apiserver with the feature gate on: ```sudo ./kube-apiserver --etcd-servers=http://127.0.0.1:2379 --feature-gates=PerNetworkServiceIPAlloc=true```
3. create network CRD by ```kubectl create -f <repo-root>/pkg/controller/artifacts/crd-network.yaml```
4. create network test2 using yaml
```yaml
apiVersion: "arktos.futurewei.com/v1"
kind: Network
metadata:
  name: test2
spec:
  type: flat
  service:
    ipam: Arktos
    cidrs: [192.168.100.0/24]
```
5. create a service associated to test2 network, using yaml
```yaml
apiVersion: v1
kind: Service
metadata:
  name: my-nginx-svc
  labels:
    run: my-nginx
    arktos.futurewei.com/network: test2
spec:
  ports:
  - port: 80
    protocol: TCP
  selector:
    run: my-nginx
```
6.verify the service also gets IP address from k8s pool
```
$ ./cluster/kubectl.sh get svc my-nginx-svc
NAME           TYPE        CLUSTER-IP   EXTERNAL-IP   PORT(S)   AGE
my-nginx-svc   ClusterIP   192.168.100.68   <none>        80/TCP    2s
``` 

**Does this PR introduce a user-facing change?**:
Partial: No by default; Yes potentially.

```release-note
If kube-apiserver start with --feature-gates=PerNetworkServiceIPAlloc=true, service IP address shall be allocated from network service IP pool.
```
